text/0000-natvis.md

```diff
@@ -0,0 +1,428 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC aims to improve the debugging experience for Rust developers, by
+enabling Rust developers to package debugger visualization scripts with their
+crates.
+
+# Motivation
+[motivation]: #motivation
+
+Most, if not all, Rust developers will at some point have to debug an issue
+in their crate. Trying to view types as they are laid out in memory is not
+always the most telling. Furthermore when viewing types from external crates,
+the information is even harder to interpret. 
+
+Many languages and debuggers enable developers to control how a type is
+displayed in a debugger. These are called "debugger visualizations" or "debugger
+views". Debugger views are merely a convenience for some types, such as
+`Vec<T>`, but are essential for types such as `HashMap<T>`, where non-trivial
+logic is needed in order to correctly display the contents of a type.
+
+Currently, Rust provides visualizations for a handful of types defined in its
+standard library via `.natvis` files or python scripts. However, this support
+is inflexible; updating it requires modifying the Rust toolchain itself,
+and either using a local build of the toolchain or waiting for a new upstream
+build of the toolchain. It is not feasible for developers of ordinary crates
+to update the Rust toolchain, solely to add visualizations for their crates.
+
+The expected outcome of this RFC is to design a way for developers to
+seamlessly integrate debugger visualizations with their crates. This would mean:
+
+* Any developer can add debugger visualizations to their crate.
+* If a Rust developer uses a crate that has debugger visualizations in it,
+  then the visualizations of those external crates will "just work" when viewed
+  under a debugger without the need of any manual configuration.
+* Supports existing debugging visualization systems. We do not propose to
+  define a new debugger visualization system; that would be a tremendous
+  undertaking, and would ignore the value of existing systems.
+* No impact on code quality or size.
+* No impact on crates that do not use debugger visualizations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC explores making Natvis debugger visualizations extensible, in Rust.
+The scenario that we want to enable is:
+
+* Alice publishes a crate, say, `cool_stuff`. Alice wrote debugger
+  visualizations for `cool_stuff`, and included them in the crate.
+* Bob is writing a new Rust application. Deep in the crate dependency graph of
+  Bob's application, some crate uses `cool_stuff`.
+  (Bob is not even aware of the existence of debugger visualizations.)
+* While Bob is debugging the application, and examining data structures,
+  he comes across an instance of `cool_stuff::CoolType` in the debugger.
+  Because Rust and the debugger know about the visualizations that Alice wrote,
```

Comment:
I'm not clear on what "Rust" knowing about visualisations means here since there is no Rust runtime or Rust debugger. Presumably, it means that the Rust compiler has built-in some knowledge of the visualisations into the debuginfo in the binary? I think it would be beneficial to be precise here.

Reply:
I'll clarify this to be precise.